### PR TITLE
TensorIterator coalesce_dimensions to not coalesce reduction dim

### LINKED ISSUE #120229: Vector norm significantly slower on 1D data for CPU code
### 🐛 Describe the bug

I was running some math code in PyTorch and noticed that the linalg.norm/norm code was the bottleneck of the performance. Turns out, for one dimensional data, e.g., of shape [n,1], norm is about 50 times slower than with a manual computation. This issue does not occur in higher dimensions, e.g., [n,2], and is easily replicable within colab. This issue only occurs for CPU code, not CUDA code, and regardless of datatype.

```py 
import torch
print('Running 1D:')
# 1D
x = torch.randn(2**18, 1)
y = torch.randn(2**18, 1)
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)
x = x.cuda()
y = y.cuda()
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)

print('Running 2D:')
# 2D
x = torch.randn(2**18, 2)
y = torch.randn(2**18, 2)
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)
x = x.cuda()
y = y.cuda()
%timeit torch.sqrt((x - y).pow(2).sum(1))
%timeit torch.norm(x - y, 2, 1)```
```

Output:
```Running 1D:
1.28 ms ± 151 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
33.8 ms ± 727 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
87.2 µs ± 67 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
67.5 µs ± 122 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Running 2D:
3.43 ms ± 453 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
3.11 ms ± 879 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
59.9 µs ± 262 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
47.9 µs ± 104 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)```

### Versions

Collecting environment information...
PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.30GHz
CPU family:                         6
Model:                              63
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           4599.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           256 KiB (1 instance)
L3 cache:                           45 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable; SMT Host state unknown
Vulnerability Meltdown:             Vulnerable
Vulnerability Mmio stale data:      Vulnerable
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.1.0
[conda] Could not collect

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jianyuh @nikitaved @pearu @mruberry @walterddr @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by min-jean-cho**
Fixes https://github.com/pytorch/pytorch/issues/120229

Performance regression with `torch.norm(input, p=2, dim=-1)` when input tensor is of shape `[*, 1]` was reproducible.

The root cause is that for an input tensor of shape `[*, 1]`, `norm_kernel_tensor_iterator_impl` goes to non-vectorized norm-2 path even though reducing over the last dim (`dim=-1`) (i.e., most inner dim). And this is due to `coalesce_dimensions` coalescing any two adjacent dims if either dim is of size 1, even if the dim is the dimension being reduced. 
I came across 
- https://github.com/pytorch/pytorch/pull/47642
- https://github.com/pytorch/pytorch/pull/47237

I am not sure of the background why it had been backed out.

Another method is to reformulate norm-p, `torch.norm(input, p, dim)`, when `input.shape[dim] == 1` for any int `p` and int `dim` as norm-p across size 1 simplifies. This method may observe better performance. 

**Performance**
Avg Latency (ms) of `torch.norm(input, p=2, dim=-1)`
Tested on 28 physical cores/socket, 1 socket on Skylake.
 
|                 	| **Avg Latency (ms)**                 	|           	|                                  	|
|-----------------	|--------------------------------------	|-----------	|----------------------------------	|
| **input shape** 	| **Before (original implementation)** 	| **After** 	| **Speedup Ratio (Before/After)** 	|
| (2**18, 1)      	| 34.00740147                          	| 0.23725   	| 143.3397447                      	|
| (2**18, 2, 1)   	| 67.57253408                          	| 0.237517  	| 284.4951416                      	|

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/TensorIterator.cpp [hunk 1]
```diff
@@ -643,6 +643,14 @@ void TensorIteratorBase::coalesce_dimensions() {
   // We can coalesce two adjacent dimensions if either dim has size 1 or if:
   // shape[n] * stride[n] == stride[n + 1].
   auto can_coalesce = [&](int dim0, int dim1) {
+    if (is_reduction_) {
+      for (auto& op : operands_) {
+        if (op.is_output && (op.stride_bytes[dim0] == 0 || op.stride_bytes[dim1] == 0 )) {
+          return false;
```
Comment (jgong5) [inline aten/src/ATen/TensorIterator.cpp:649]: Would this impact the case where we reduce on two dims for which we can actually coalesce?

File: aten/src/ATen/TensorIterator.cpp [hunk 2]
```diff
@@ -643,6 +643,14 @@ void TensorIteratorBase::coalesce_dimensions() {
   // We can coalesce two adjacent dimensions if either dim has size 1 or if:
   // shape[n] * stride[n] == stride[n + 1].
```
Comment (jgong5) [inline aten/src/ATen/TensorIterator.cpp:644]: Need to update the comment accordingly?

**Review by jgong5 — CHANGES_REQUESTED** [review]
Seems tests are failing...

File: aten/src/ATen/TensorIterator.cpp (hunk 1 above)
Comment (min-jean-cho) [reply aten/src/ATen/TensorIterator.cpp:649]: I see some tests are failing due to precision issue. I think it's better to re-formulate p-norm when `input.shape[dim] == 1`.

**Comment by min-jean-cho** [comment]
Moved to https://github.com/pytorch/pytorch/pull/122143. Simplifying Lp norm formula for 1-dimensional vector.
